### PR TITLE
[Profiler] Profiling support for parallel programs

### PR DESCRIPTION
**Pull request by ayakayorihiro**
This PR contains:
- The `profiler-instrumentation` pass is now updated to contain four types of probes in order to capture call site information.
  - `group_probe`: A probe that is high when a group is active.
  - `primitive_probe`: A probe that is high when a primitive cell is active.
  - `se_probe`: A probe that is high when a group that structurally enables another group is active.
  - `cell_probe`: A probe that is high when a group that invokes a non-primitive cell is active.
- `profiler-process.py`: an updated VCD postprocessing script that generates `.folded` and `.dot` files:
  - A "flattened" flame graph, where each parallel arm gets its own "cycle". So, if arm A and arm B were executing on a single cycle, the flame graph would account for a cycle in arm A and a cycle in arm B.
  - A scaled flame graph, where a cycle is divided between the parallel arms in execution. So, if arm A and arm B were executing on a single cycle, the flame graph would account for 0.5 cycles in arm A and 0.5 cycles in arm B.
  - `aggregate.dot`: A tree summary of the execution of the program. Nodes (groups and cells) are labeled with the number of times the node was a leaf, and edges are labeled with the number of cycles that edge was activated.
  - `rank{i}.dot`: A tree representation of the `i`th most active stack picture. `rankings.csv` lists the specific cycles that each ranked tree was active for.
- Updated fud2 support (`fud2/scripts/profiler.rhai`)
- Utility scripts to help convert generated `.folded` and `.dot` files into visualizations (`svg` and `png` files respectively)
- Updated tests/cleanup of old tests that relied on scripts that no longer exist

As usual, I'd really appreciate any suggestions or thoughts!!

### CODE REVIEW

File: docs/running-calyx/profiler.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+*Note: The profiler is still in development. If you have any suggestions, thoughts, or feedback, please let us know!*
+
+Profilers can help you analyze performance information to find places you can optimize in your code. Currently, the Calyx profiler prouduces timing information in terms of cycle counts.
+
+## Setup
+
+In order to run the profiler, you need:
+- [fud2][]
+- The Python `vcdvcd` library
```
Comment (ekiwi) [inline docs/running-calyx/profiler.md:11]: how are people supposed to get this library? could `fud2` automatically install it in the virtual environment it creates?

File: tools/profiler/finagle-with-svg.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import os
+import sys
+
+# Takes in a flame graph svg that is scaled by 1000 and prints a version with fixed cycles.
+def main(svg_in):
```
Comment (ekiwi) [inline tools/profiler/finagle-with-svg.py:5]: maybe you can come up with a better name for this script that contains a bit of the description?

**Review by ekiwi — CHANGES_REQUESTED** [review]
Two small changes and then we can merge this in.

File: docs/running-calyx/profiler.md (hunk 1 above)
Comment (ayakayorihiro) [reply docs/running-calyx/profiler.md]: I linked to the repo and suggested an installation command, how does the new bullet look?

File: tools/profiler/finagle-with-svg.py (hunk 2 above)
Comment (ayakayorihiro) [reply tools/profiler/finagle-with-svg.py]: Changed to `adjust-scaled-flame-svg.py`!

File: docs/running-calyx/profiler.md (hunk 1 above)
Comment (ekiwi) [reply docs/running-calyx/profiler.md]: @EclecticGriffin do you know if there is a way to have this python library installed when running `fud2 env init`? I think that would be the most straightforward way for people to get this dependency.

Comment (ayakayorihiro) [reply docs/running-calyx/profiler.md]: Ah I didn't know about `fud2 env init`, I'll add `vcdvcd` to the init commands!

Comment (ayakayorihiro) [reply docs/running-calyx/profiler.md]: Added the install command here: https://github.com/calyxir/calyx/pull/2370/files#diff-8f6eee891971328771ea1488c38a76b5722005c2e857008c8289fc223f0040b9

Comment (ekiwi) [reply docs/running-calyx/profiler.md]: That is great, thanks!

**Review by ekiwi — APPROVED** [review]
:shipit: !

@ayakayorihiro : is this ready to be merged?

**Comment by ayakayorihiro** [comment]
Merging :)

**Pull request merged**